### PR TITLE
volume: use AtomicWriteFile to save volume options

### PR DESCRIPTION
**Pull request by cyphar**
If the system (or Docker) crashes while saivng the volume options, on restart the daemon will error out when trying to read the options file because it doesn't contain valid JSON.

In such a crash scenario, the new volume will be treated as though it has the default options configuration. This is not ideal, but volumes created on very old Docker versions (pre-1.11[1], circa 2016) do not have opts.json and so doing some kind of cleanup when loading the volume store (even if we take care to only delete empty volumes) could delete existing volumes carried over from very old Docker versions that users would not expect to disappear.

Ultimately, if a user creates a volume and the system crashes, a volume that has the wrong config is better than Docker not being able to start.

1: commit b05b2370757d ("Support mount opts for `local` volume driver")

**- Description for the changelog**
```markdown changelog
daemon: volume: write options JSON atomically to avoid "invalid JSON" errors after system crash
```

### CODE REVIEW

**Comment by cyphar** [comment]
I would like to have some kind of cleanup code during `New` to clear these broken volumes, but it seems that pre-1.11 Docker volumes are missing `opts.json` too (hence the code to support loading them).

Another option would be to create a directory tree with the `opts.json` and move it into place atomically, but that would require adding some new bits to `AtomicWriteSet` and I suspect it would require extra work to make sure it is correct on Windows.

**Comment by thaJeztah** [comment]
Is that about the `_data` directory, and old versions not having that, or the `opts.json` itself? For the `_data` directory, I think we did a migration, or at least these were related to that: code for upgrading was added in 4ad05ed9859614203708eec3d3a1ecf0cc81e17a (https://github.com/moby/moby/pull/13728) (with a follow-up for downgrading in 0abd7ba22998e2ad00091c4ddc0f9ea28625adfe (https://github.com/moby/moby/pull/43591)). I must add here that we always made migration code with the expectation of users not skipping multiple versions (so upgrading from a very old version of docker to "latest" would not have a promise of all intermediate migrations to be handled). Docker 1.11 is really old, and we no longer support API versions related to it (minimum API version is now 1.24, which docker 1.12)

As to the `opts.json` being present in itself; I think it only gets created if options were _passed_ when creating the volume, and no file created otherwise (at least on `volume create`), so indeed, it may be an expected situation for the file to not be present.

File: volume/local/local.go [hunk 1]
```diff
@@ -388,7 +389,7 @@ func (v *localVolume) saveOpts() error {
 	if err != nil {
 		return err
 	}
-	err = os.WriteFile(filepath.Join(v.rootPath, "opts.json"), b, 0o600)
+	err = ioutils.AtomicWriteFile(filepath.Join(v.rootPath, "opts.json"), b, 0o600)
```
Comment (thaJeztah) [inline volume/local/local.go:392]: I had to double-check if using this here could cause issues if `/var/lib/docker/volumes` was on a different filesystem (I know that's a scenario I've seen, where that path was mounted from a different device), but it looks like it uses the current directory as path for the temporary files it creates, so I think we're good; https://github.com/moby/moby/blob/11179de64c6e4a1f770caae2c975911077044be7/pkg/ioutils/fswriters.go#L14-L17

**Review by thaJeztah — APPROVED** [review]
LGTM

but perhaps @cpuguy83 @vvoland want to have a look as well 👍

**Comment by cyphar** [comment]
Ah yeah, I missed that `setOpts` won't call `saveOpts` if there are no options.

In that case, the only way to avoid these volumes sticking around in the crash case is to create the tree somewhere else and atomically rename it into place. Since volume names cannot start with `.` we could create directories called `.new-volume-VOLUME_NAME.XXXXXX` and delete them during `New` when scanning?

**Comment by thaJeztah** [comment]
Perhaps that would work yes. I think that'd be fine for a separate PR to look into. Or at least, from what I can see the changes in this PR are an OK improvement on its own, and it's very scoped (making it a possible candidate for backporting where needed).

For the proposed idea of `Create` being more atomic as a whole (so to prevent partial volumes being left behind), we'd need to look at best options.

- 👍 The proposed solution would be for `local.New()` to check if temp-locations (`.new-volume-VOLUME_NAME.XXXXXX`) exist, and to delete those (`local.New()` is called from `NewVolumeService` through `setupDefaultDriver`); that happens during daemon startup, so indeed seems like an ok moment to handle this.
- ❓ A possible alternative could be to postpone creating the `_data` directory until after we've successfully wrote the opts; currently we create that directory _before_ we write the options; https://github.com/moby/moby/blob/11179de64c6e4a1f770caae2c975911077044be7/volume/local/local.go#L163-L166
- ☝️ If we performed that step _after_, we could detect that the volume creation didn't complete. In addition; a volume without `_data` also would be relatively "safe" to delete (no data lost).


Somewhat orthogonal, but looking at the existing code, I'm actually wondering if this whole chunk should've been a `createVolume()` constructor 🤔 ; https://github.com/moby/moby/blob/11179de64c6e4a1f770caae2c975911077044be7/volume/local/local.go#L150-L177

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Pull request merged**